src/ripple/app/paths/impl/BookStep.cpp

```diff
@@ -106,6 +109,12 @@ class BookStep : public StepImp<TIn, TOut, BookStep<TIn, TOut>>
         return EitherAmount (cache_->out);
     }
 
+    bool
+    redeems (ReadView const& sb, bool fwd) const override
+    {
+        return !ownerPaysTransferFee_;
```

Comment:
For DirectSteps, we need to look in the ledger. For BookSteps, unless the offer owner is also the issuer, then the owner is redeeming. However, when the owner pays the transfer fee, we set `redeems` to false so the next step doesn't _also_ charge the fee (double charging). When the sender pays we set it to true and let the next step do the right thing with the transfer fee.

Note I did not handle the complexity of when the issuer is the owner.